ci/generate_docs.py

```diff
@@ -0,0 +1,219 @@
+                            all_jobs[job_name] = {
+                                "description": comment_buffer,
+                                "line_no": index + 1,
+                                "script": gitlab_file_content[job_name].get(
+                                    "script",
+                                    ["No script defined, probably extends another job"],
+                                ),
+                            }
+                    comment_buffer = []
+
+        return all_jobs
+
+    def save_docs_into_file(self) -> None:
+        """Dump all the information into a documentation file."""
+        with open(self.DOC_FILE, "w") as doc_file:
+            # Some general info for the whole CI
+            doc_file.write("# CI pipeline\n")
+            doc_file.write(
+                "It consists of multiple stages below, each having one or more jobs\n"
+            )
+            latest_master = "https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/master/latest"
+            doc_file.write(
+                f"Latest CI pipeline of master branch can be seen at [{latest_master}]({latest_master})\n"
+            )
+
+            # TODO: test-hw are run inside test stage, maybe unite it under it
+            for file, file_info in self.ALL_JOBS.items():
+                # Generating header with a link to the file
+                doc_file.write(
+                    f"## {Path(file).stem.upper()} stage - [file](../{file})\n\n"
+                )
+
+                # TODO: are we alright using this python >= 3.8 feature (walrus operator)?
+                if description := file_info["overall_description"]:
+                    for line in description:
+                        doc_file.write(f"{line}\n")
+                    doc_file.write("\n")
+
+                job_amount = f"{len(file_info['jobs'])} job{'s' if len(file_info['jobs']) > 1 else ''}"
+                doc_file.write(f"Consists of **{job_amount}** below:\n")
+
+                for job_name, job_info in file_info["jobs"].items():
+                    # Generating smaller header with link to the exact line of
+                    # this job in the master branch
+                    # (will work properly only after merging changes to master)
+                    github_job_link = f"https://github.com/trezor/trezor-firmware/blob/master/{file}#L{job_info['line_no']}"
+                    doc_file.write(f"- ### [{job_name}]({github_job_link})\n")
+                    if not job_info["description"]:
+                        doc_file.write("Missing description\n")
+                    for line in job_info["description"]:
+                        doc_file.write(f"{line}\n")
+
+                    # Code block for the whole script
+                    # TODO: for longer scripts, we could make a collapsible section
+                    # https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab
```

Comment:
Not sure if we should include the script at all - with the nested nix-shells and poetry it will probably confuse the reader. They can open the yml file easily if they want the definition.